Ambit/Source/Ambit/Mode/ConfigImportExport.spec.cpp

```diff
@@ -1017,4 +1017,101 @@ void ConfigImportExportSpec::Define()
             JsonContent = nullptr;
         });
     });
+
+    Describe("OnExportMap()", [this]()
+    {
+        BeforeEach([this]()
+        {
+            // Create Mode Settings
+            GLevelEditorModeTools().ActivateMode(FAmbitMode::EM_AmbitModeId);
+
+            FAmbitMode* AmbitMode = FAmbitMode::GetEditorMode();
+            check(AmbitMode);
+
+            // Set AWS S3 bucket to store test data
+            AmbitMode->UISettings->AwsRegion = "us-west-2";
+            AmbitMode->UISettings->S3BucketName = "ambit-unit-test-data";
+
+            // Set target platform for export
+            AmbitMode->UISettings->ExportPlatforms.SetWindows(true);
+
+            // Generate World
+            World = FAutomationEditorCommonUtils::CreateNewMap();
+
+            // Generate Exporter
+            Exporter = NewObject<UMockableConfigImportExport>();
+        });
+
+        It("Should Return FReply::Handled", [this]()
```

Comment:
Future consideration: This test covers the "happy path". Are there any conditions (S3 service problems, disk write errors, etc.) that could cause a map export to fail? If so, we should add test coverage to validate the intended behavior of our code if/when those conditions occur. This requires using mocking to emulate failures of the APIs we're talking to.

No change needed for this PR, though.

Reply:
Any other conditions would be out of bounds for this test. For example, S3 service problems should be a separate suite of tests, especially since the actual S3 calls have been replaced with mock S3 calls, so it doesn't make any _real_ calls into the service.

Reply:
I agree it would require additional tests. But just to make sure we're on the same page, if our code is intended to react to rare but possible failures of collaborating services—for example, to display a UE error dialog to the user or to log a contextual error message–then we should write tests that validate that behavior. This would typically be done using a special mock sometimes referred to as a "saboteur". But like I said, this is merely a suggestion for improving future test coverage. Nothing to hold up this PR on.

Reply:
I agree with Kris. While there are other functions that may have problems, if our code actions/does not action properly on their inability to perform properly (and it generally should, we need test cases to match the expectation of what they failed with to ensure that their failure is consistently tracked for failing. This also allows us a time to reflect if our main code can be improved by going through and thinking through the tests that it needs. 

For example, on line [582](https://github.com/aws-samples/aws-ambit-scenario-designer-ue4/blob/main/Ambit/Source/Ambit/Mode/ConfigImportExport.cpp#L582), we are writing to disk. That function throws exceptions in critical failure. How does our function (`OnExportGltf()`) react to that failure? Should it react better to that failure? If so, this is the opportunity to go back and make that modification, strengthening not only our testing suite (by pseudo TDD), but the actual code it is meant to protect. 

I personally think there is a lot of value of adding the unhappy test cases for the current iteration if we have time.

Reply:
Now that I look at this test, there are more concerns than just the fact that it tests a "happy path". **This specific test on it's own is fine**, but I would expect there to be more tests in this "OnExportMap()" group.

 **I don't think we have to consider this in scope for this PR, but know that the current test coverage for this method isn't as robust as it should be.**

Here are some examples of additional tests I would expect (using lots of mocks to intercept and validate appropriate calls)...

`"should save packaged map to local director X"`

`"should upload file to S3"` _👈 This test would confirm—via mocking—that the S3 client is called properly with correct parameters_

- `"given target platform 'Linux'..."`
    - `"should package a Linux map"` _👈 You would mock the FPlatformProcess::CreateProc() method to confirm it gets called with the proper args_
    - `"should compress packaged map as tar.gz"`

- `"given target platform 'Win64'..."`
    - `"should package a Windows map"` _👈 Same as above_
    - `"should compress packaged map as ZIP"`

Then there is the failure handling that should be tested as discussed in previous comments. Example:

`"when error occurs during packaging, it should log failure message"`

Reply:
I agree with both comments. We do need more test coverage for these specific aspects.